[feat] login/join

gyur1kim opened the pull request:
### 진행사항


- [x] 아이디/비번 입력 틀리면 error message
- [x] 로그인 화면에서 회원가입/비밀번호 찾기 화면 이동 버튼
- [x] 네비게이션 스택 적용
- [x] 비밀번호 입력 시 보였다 안보였다 하는 기능
- [x] 로그인 성공 시 home screen으로 이동
- [x] api문서에 확인 후 data params명 적용

---

### 추후 해야할 일
- [ ] password 규칙 적용
- [ ] authentication
- [ ] 토큰 유효성 검사
- [ ] be와 api 테스트
- [ ] 웹뷰와 쿠키 공유
  
등등등..

Closes issue #4

gyur1kim committed:
feat: useController를 적용한 Input component 생성

- name props로 react-hook-form에 등록하는 것 같다
- name과 control을 받아서 자동으로 onChange, onBlur, value 값을 지정

gyur1kim committed:
feat: 아이디/비밀번호 없을 시 에러메시지 띄우기

- rules: required이용
- 추후 이메일 정규식, 최대길이 등 제한 가능
- FormInput의 타입을 받을 때 UseControllerProps를 이용
- ErrorMessage 패키지 설치

gyur1kim committed:
feat: 쿠키를 이용하여 로그인 기능 구현

gyur1kim committed:
chore: 임시커밋

- 내비게이션 하는 중

gyur1kim committed:
style: 파일 이름 변경

gyur1kim committed:
feat: login화면에서 비밀번호 찾기, 회원가입 페이지 이동

- native-stack에서 커스텀이 자유로운 stack으로 변경
- page params를 상수로 관리

gyur1kim committed:
feat: 회원가입, 비밀번호 찾기 기능

gyur1kim committed:
feat: radio button 생성

gyur1kim committed:
refactor: navigator 관리 상수 및 타입 하나로 합치기

gyur1kim committed:
fix: navigation stack 수정

- 기존의 '로그인 화면으로 돌아가기' 버튼은 login > signup > login 방식으로 쌓임
- pop(1)을 통해 login 하나만 유지

gyur1kim committed:
style: id를 email로 변경

devysi0827 committed:
[doc] fastlane doc update

devysi0827 committed:
[doc] pr template update

gyur1kim committed:
feat: join api 적용

gyur1kim committed:
style: signup -> join으로 이름 변경

gyur1kim committed:
feat: password 입력 시 입력한 값 visibility 토글 기능 추가

gyur1kim committed:
feat: 로그인 성공 시 home 화면으로 이동

gyur1kim committed:
feat: 로그아웃 버튼 추가

devysi0827 commented on app/rootLayout.tsx:
```diff
@@ -1,12 +1,29 @@
   return (
-    <Stack.Navigator initialRouteName="Login">
-      <Stack.Screen name="Login" component={LoginScreen} />
+    // <Stack.Navigator initialRouteName={navigatorParams.HOME}>
```
불필요한 주석은 지우고 올리는 게 좋습니다! (아래 많은데 생략할게요)

devysi0827 commented on app/shared/components/Input/Input.tsx:
```diff
@@ -15,6 +33,126 @@ const Input: React.FC<InputProps> = ({ label, ...props }) => {
+      </View>
+    </>
+  );
+};
```
코드 중복이 너무 심해요. 개선하시는 게 좋아보입니다.

why :
- 비슷한 상황이 생기면 Input 컴포넌트가 지속적으로 늘어난다. (formInput, queryInput, ...)
- 이 때 Input의 디자인이 바뀌거나 사용성 개선 요청이 들어온다면, n개의 컴포넌트 동시 수정해야함. QA든 개발자든 일이 n배여서 이는 보통 하나만 수정하는 게 효율적임.

solution : 
- 책임을 잘 분리해보세요. 
  - Input은 입력을 받고 보여주는 역할(UI)을 수행합니다.
  - FormInput은 Input과 데이터를 react-hook-form에서 불러오는 두 가지 역할을 수행합니다. 그래서 중복이 생깁니다.
  - Input이 UI 수정만을 담당하고, FormInput이 데이터를 가져오는 역할만 수행한다면,  why에 두 번째줄은 Input만 수정하면 해결되는 문제입니다.

gyur1kim replied:
그냥 input 컴포넌트 하나로 합쳤습니다..!
굳이 나누는게 더 비효율적이라고 판단했습니다.

- 기존의 useController로 따로 제작한 form전용 input 컴포넌트들을 없앴습니다.
- Controller를 이용해 Input 컴포넌트를 render합니다.
- render내에서 onBlur 등을 처리합니다.

devysi0827 commented on app/screens/ForgotPasswordScreen.tsx:
```diff
@@ -0,0 +1,89 @@
+
+  const onSubmit = async (data: ForgotPasswordForm) => {
+    try {
+      await axiosInstance.post('/api/forgot-password', { id: data.id });
```
[ 기획 ]
이 부분 논의 안된 거 같은 데, 나중에 정식으로 이야기해봐요!

=> 이메일로 보낼 지, password 수정 창을 사용할 지
=> password 수정 창은 결국 필요하네요

devysi0827 commented on app/screens/HomeScreen.tsx:
```diff
@@ -0,0 +1,14 @@
+  return (
+    <View>
+      <Text>로그인 되고 나올 화면</Text>
+      <Button title="로그아웃" onPress={() => logout()} />
```
[ 퀴즈 ]
해당 줄을 개선해야할 필욕 있을까요?

devysi0827 commented on app/screens/JoinScreen.tsx:
```diff
@@ -0,0 +1,191 @@
+        });
+      }, 2000);
+    } catch (error) {
+      setMessage('회원가입 실패');
```
실패한 사유가 나중에 세분화되면 좋을 거 같네용

devysi0827 commented on app/shared/components/RadioButton/RadioButton.tsx:
```diff
@@ -0,0 +1,104 @@
+  );
+};
+
+const FormRadioButton = <T extends FieldValues>({
```
위에 Input과 같은 문제입니다 : 이하 생략

devysi0827 commented on app/shared/utils/axiosInstance.ts:
```diff
@@ -1,33 +1,49 @@
+        return axiosInstance(originalRequest);
+      } catch (error) {
+        // 리프레시 토큰도 만료된 경우
+        useAuthStore.getState().logout();
```
여기서 popup 창 "로그인이 만료되었습니다"도 나오면 너무 좋을 거 같아요!! 
이게 axios.instance에서 처리해도 되는 지 모르겠네요

gyur1kim replied:
좋은 생각입니다!!
나중에 화면 디자인 나오면 추가적으로 고민해보겠습니다.

gyur1kim committed:
fix: 불필요한 주석 삭제

gyur1kim committed:
refactor: onPress에 logout 함수 전달 수정

- logout 함수를 전달

gyur1kim committed:
refactor: native-stack 변경

- stack vs native-stack
- [native-stack] 성능이 중요하고 기본 네이티브 동작만으로 충분한 경우에 적합
- [stack] 세밀한 커스터마이징이 필요하거나, 특별한 전환 효과를 구현해야 할 때 적합
- React Navigation 공식 문서에서도 가능하면 native-stack을 사용하도록 권장
- 커스터마이징이 필요한 경우에만 stack을 사용하도록 권장

gyur1kim committed:
feat: 에러 메시지 constant로 분리

gyur1kim committed:
feat: 로그인 기능 서버 연동

- env파일 생성

gyur1kim committed:
refactor: input컴포넌트 통일

- 기존의 useController로 따로 제작한 form전용 input 컴포넌트들을 없앴습니다.
- Controller를 이용해 Input 컴포넌트를 render합니다
- render내에서 onBlur 등을 처리합니다.

devysi0827 commented on app/shared/constants/message.ts:
```diff
@@ -0,0 +1,43 @@
+export const loginMessage = {
```
(여기 처음 남겨보는데... 잘 보이겠죠...?)

loginMessage를 LOGINMESSAGE로 표현하는 것도 좋아보여요!
- airbnb 스타일 가이드인데, 변하지 않는 상수값을 대문자로 표현하면 애는 찐 상수구나 생각이 들어서 유지보수 시, 좋았습니다.

https://github.com/airbnb/javascript/#naming--uppercase

gyur1kim replied:
수정하였습니다.

머지해도 되겠죠....?

devysi0827 replied:
그럼요 merge 버튼 누르란 말을 빼먹었네요

gyur1kim committed:
refactor: 비밀번호 찾기 페이지 Input 컴포넌트 반영

gyur1kim committed:
style: 상수 데이터 대문자 표기
